File: JSCodes/src/communication/RealtimeHandler.js
const WebSocket = require('ws');
const SecureProtocol = require('./SecureProtocol');
const { ErrorHandler, ErrorSeverity } = require('../utils/ErrorHandling');
const ConfigManager = require('../utils/ConfigManager');

class RealtimeHandler {
    constructor() {
        this.secureProtocol = new SecureProtocol();
        this.connections = new Map();
        this.messageHandlers = new Map();
        this.heartbeatInterval = null;
        this.config = ConfigManager.getInstance();
    }

    async startServer(host = 'localhost', port = 8765) {
        try {
            const server = new WebSocket.Server({ host, port });
            this.startHeartbeat();

            server.on('connection', this.handleConnection.bind(this));
            return server;
        } catch (error) {
            throw ErrorHandler.createError(
                `Failed to start realtime server: ${error}`,
                ErrorSeverity.ERROR,
                'REALTIME_HANDLER'
            );
        }
    }

    async handleConnection(websocket) {
        try {
            const sessionId = await this.secureProtocol.establishSecureChannel(websocket);
            this.connections.set(sessionId, websocket);

            websocket.on('message', async (message) => {
                await this.processMessage(sessionId, message);
            });

            websocket.on('close', () => {
                this.handleDisconnection(sessionId);
            });
        } catch (error) {
            throw ErrorHandler.createError(
                `Connection handling failed: ${error}`,
                ErrorSeverity.ERROR,
                'REALTIME_HANDLER'
            );
        }
    }

    async processMessage(sessionId, message) {
        try {
            const decrypted = await this.secureProtocol.decryptMessage(sessionId, message);
            const messageType = decrypted.type;

            if (this.messageHandlers.has(messageType)) {
                const handlers = this.messageHandlers.get(messageType);
                for (const handler of handlers) {
                    await handler(sessionId, decrypted);
                }
            }
        } catch (error) {
            throw ErrorHandler.createError(
                `Message processing failed: ${error}`,
                ErrorSeverity.ERROR,
                'REALTIME_HANDLER'
            );
        }
    }

    handleDisconnection(sessionId) {
        if (this.connections.has(sessionId)) {
            this.connections.delete(sessionId);
        }
        this.secureProtocol.closeSession(sessionId);
    }

    startHeartbeat() {
        this.heartbeatInterval = setInterval(async () => {
            try {
                for (const [sessionId, websocket] of this.connections) {
                    try {
                        if (websocket.readyState === WebSocket.OPEN) {
                            await this.sendPing(websocket);
                        } else {
                            this.handleDisconnection(sessionId);
                        }
                    } catch {
                        this.handleDisconnection(sessionId);
                    }
                }
            } catch (error) {
                throw ErrorHandler.createError(
                    `Heartbeat failed: ${error}`,
                    ErrorSeverity.ERROR,
                    'REALTIME_HANDLER'
                );
            }
        }, 30000); // 30 seconds interval
    }

    async sendPing(websocket) {
        return new Promise((resolve, reject) => {
            websocket.ping((error) => {
                if (error) reject(error);
                else resolve();
            });
        });
    }

    registerHandler(messageType, handler) {
        if (!this.messageHandlers.has(messageType)) {
            this.messageHandlers.set(messageType, new Set());
        }
        this.messageHandlers.get(messageType).add(handler);
    }

    async broadcast(message) {
        try {
            const promises = Array.from(this.connections.entries()).map(
                async ([sessionId, websocket]) => {
                    if (websocket.readyState === WebSocket.OPEN) {
                        const encrypted = await this.secureProtocol.encryptMessage(sessionId, message);
                        websocket.send(encrypted);
                    }
                }
            );
            await Promise.all(promises);
        } catch (error) {
            throw ErrorHandler.createError(
                `Broadcast failed: ${error}`,
                ErrorSeverity.ERROR,
                'REALTIME_HANDLER'
            );
        }
    }

    stop() {
        if (this.heartbeatInterval) {
            clearInterval(this.heartbeatInterval);
            this.heartbeatInterval = null;
        }

        for (const [sessionId, websocket] of this.connections) {
            websocket.close();
            this.handleDisconnection(sessionId);
        }
    }
}

module.exports = RealtimeHandler; 

File: PythonCodes/src/communication/realtime_handler.py
import asyncio
import websockets
from typing import Dict, Set, Callable, Any
from .secure_protocol import SecureProtocol
from ..utils.error_handling import KAPESITError, ErrorSeverity
from ..utils.config_manager import config_manager

class RealtimeHandler:
    def __init__(self):
        self.secure_protocol = SecureProtocol()
        self.connections: Dict[str, websockets.WebSocketServerProtocol] = {}
        self.message_handlers: Dict[str, Set[Callable]] = {}
        self.heartbeat_task = None
        
    async def start_server(self, host: str = "localhost", port: int = 8765):
        try:
            self.heartbeat_task = asyncio.create_task(self._heartbeat_loop())
            server = await websockets.serve(self._handle_connection, host, port)
            return server
        except Exception as e:
            raise KAPESITError(
                f"Failed to start realtime server: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="REALTIME_HANDLER"
            )
    
    async def _handle_connection(self, websocket: websockets.WebSocketServerProtocol, path: str):
        try:
            session_id = await self.secure_protocol.establish_secure_channel(websocket)
            self.connections[session_id] = websocket
            
            try:
                async for message in websocket:
                    await self._process_message(session_id, message)
            except websockets.exceptions.ConnectionClosed:
                await self._handle_disconnection(session_id)
                
        except Exception as e:
            raise KAPESITError(
                f"Connection handling failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="REALTIME_HANDLER"
            )
    
    async def _process_message(self, session_id: str, message: bytes):
        try:
            decrypted = await self.secure_protocol.decrypt_message(session_id, message)
            message_type = decrypted.get("type")
            
            if message_type in self.message_handlers:
                for handler in self.message_handlers[message_type]:
                    await handler(session_id, decrypted)
                    
        except Exception as e:
            raise KAPESITError(
                f"Message processing failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="REALTIME_HANDLER"
            )
    
    async def _handle_disconnection(self, session_id: str):
        if session_id in self.connections:
            del self.connections[session_id]
        self.secure_protocol.close_session(session_id)
    
    async def _heartbeat_loop(self):
        while True:
            try:
                for session_id, websocket in self.connections.items():
                    try:
                        await websocket.ping()
                    except:
                        await self._handle_disconnection(session_id)
                await asyncio.sleep(30)
            except Exception as e:
                raise KAPESITError(
                    f"Heartbeat loop failed: {str(e)}",
                    severity=ErrorSeverity.ERROR,
                    component="REALTIME_HANDLER"
                )
    
    def register_handler(self, message_type: str, handler: Callable):
        if message_type not in self.message_handlers:
            self.message_handlers[message_type] = set()
        self.message_handlers[message_type].add(handler)
    
    async def broadcast(self, message: Dict[str, Any]):
        try:
            for session_id, websocket in self.connections.items():
                encrypted = await self.secure_protocol.encrypt_message(session_id, message)
                await websocket.send(encrypted)
        except Exception as e:
            raise KAPESITError(
                f"Broadcast failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="REALTIME_HANDLER"
            ) 